sgl-kernel/csrc/cpu/gemm_fp8.cpp

```diff
@@ -0,0 +1,356 @@
+      float* __restrict__ Ctmp,
+      const float* __restrict__ bias,
+      const float* __restrict__ scale,
+      int M,
+      int N,
+      int K,
+      int lda,
+      int ldb,
+      int ldc,
+      int64_t blocks_k_per_group) {
+    constexpr int BLOCK_N = block_size_n();
+
+    // [BLOCK_K, BLOCK_N] -> [BLOCK_K / 2, BLOCK_N * 2]
+    const int ldb_tmp = block_size_n();
+
+    // accumulate across K per BLOCK_K
+    for (int k = 0; k < K; k += BLOCK_K) {
+      int kb_size = std::min(BLOCK_K, K - k);
+
+      int idx = (k / BLOCK_K) / blocks_k_per_group;
+      unpack_B(Btmp, B + k * ldb, N, kb_size, ldb, ldb_tmp, scale[idx]);
+
+      const bool add_C = (k != 0);
+      at::native::cpublas::brgemm(
+          M, N, kb_size, lda, ldb_tmp, BLOCK_N, add_C, A + k, Btmp, Ctmp);
+    }
+
+    // copy from Ctmp to C
+    for (int m = 0; m < M; ++m) {
+      if constexpr (has_bias) {
+        copy_add_stub(C + m * ldc, Ctmp + m * BLOCK_N, bias, N);
+      } else {
+        copy_stub(C + m * ldc, Ctmp + m * BLOCK_N, N);
+      }
+    }
+  }
+};
+
+template <typename scalar_t, typename packed_t, bool has_bias>
+void tinygemm_kernel(
+    const scalar_t* __restrict__ A,
+    const packed_t* __restrict__ B,
+    scalar_t* __restrict__ C,
+    scalar_t* __restrict__ Btmp,
+    float* __restrict__ Ctmp,
+    const float* __restrict__ scale,
+    const float* __restrict__ bias,
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t lda,
+    int64_t ldb,
+    int64_t ldc,
+    bool brg,
+    int64_t blocks_k_per_group,
+    int64_t block_size_K) {
+
+  if (brg) {
+    brgemm<scalar_t, packed_t, has_bias>::apply(
+        A, B, C, Btmp, Ctmp, bias, scale, M, N, K, lda, ldb, ldc, blocks_k_per_group);
+    return;
+  }
+
+  // TODO: add the support for use_brgemm = false;
+  TORCH_CHECK(false, "use_brgemm = false is not supported yet");
+}
+
+template <typename scalar_t, typename packed_t>
+void fp8_scaled_mm_kernel_impl(
+    scalar_t* __restrict__ out,
+    const scalar_t* __restrict__ mat1,
+    const packed_t* __restrict__ mat2,
+    const float* __restrict__ scales2,
+    const float* __restrict__ bias,
+    int64_t M,
+    int64_t N,
+    int64_t K,
+    int64_t mat1_strideM,
+    int64_t out_strideM,
+    int64_t block_size_N,
+    int64_t block_size_K) {
+
+  constexpr int64_t BLOCK_M = block_size_m();
+  constexpr int64_t BLOCK_N = block_size_n();
+  const int64_t MB = div_up(M, BLOCK_M);
+  const int64_t NB = div_up(N, BLOCK_N);
+
+  const int64_t scale_size_N = div_up(N, block_size_N);
+  const int64_t scale_size_K = div_up(K, block_size_K);
+
+  const int64_t blocks_n_per_group = block_size_N / BLOCK_N;
+  const int64_t blocks_k_per_group = block_size_K / BLOCK_K;
+
+  // TODO: add the support for use_brgemm = false;
+  // use avx512-bf16 when a) M is small; b) dtype is bfloat16, otherwise use amx
+  // const bool use_brgemm = (M > 4) || (!std::is_same_v<scalar_t, at::BFloat16>);
+  const bool use_brgemm = true;
```

Comment:
```suggestion
  // use avx512-bf16 when a) M is small; b) dtype is bfloat16, otherwise use amx
  const bool use_brgemm = can_use_brgemm<at::Float8_e4m3fn>(M);
```

update in gemm.h
```
template <> inline bool can_use_brgemm<at::Float8_e4m3fn>(int M) { return false; }
```

Reply:
Code is updated in https://github.com/mingfeima/sglang/pull/47/commits/5ed43bfcf5e1ef1457471b0e24f1925858d8d540